### PR TITLE
Factor out project loading from MSBuildWorkspace

### PR DESCRIPTION
**Pull request by mattwar**
This change takes all the internal API's that MSBuildWorkspace uses to open project and solution files and presents them in a public class that produces ProjectInfo's and SolutionInfo's.instead of directly adding to the workspace.

This is useful for use with workspaces that are not MSBuildWorkspace but still wanting to pull project information from msbulid project files.

Also the API UpdateRefernecesAfterAdd has been moved to Workspace, as it is generally good for any workspace implementation.  It should probably be considered to more tightly integrate this code.

@DustinCampbell @jaredpar @Pilchie @jasonmalinowski  please review

### CODE REVIEW

File: src/Compilers/VisualBasic/VbcCore/project.lock.json [hunk 1]
```diff
@@ -1,1896 +0,0 @@
-{
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/VbcCore/project.lock.json]: Why are you deleting the lock.json files?

Comment (DustinCampbell) [inline src/Compilers/VisualBasic/VbcCore/project.lock.json]: FWIW, I noticed today that the project.lock.json file for VbcCore kept changing on my whenever I created a branch off master and rebuilt. Something may be incorrect here.

File: src/Compilers/VisualBasic/VbcCore/project.lock.json [hunk 2]
```diff
@@ -1,1896 +0,0 @@
...
-        "System.Security.Cryptography.Hashing.Algorithms.4.0.0-beta-23109.nupkg.sha512",
-        "System.Security.Cryptography.Hashing.Algorithms.nuspec",
-        "lib/DNXCore50/System.Security.Cryptography.Hashing.Algorithms.dll",
-        "lib/net46/System.Security.Cryptography.Hashing.Algorithms.dll",
-        "ref/dotnet/System.Security.Cryptography.Hashing.Algorithms.dll",
-        "ref/dotnet/System.Security.Cryptography.Hashing.Algorithms.xml",
-        "ref/dotnet/de/System.Security.Cryptography.Hashing.Algorithms.xml",
-        "ref/dotnet/es/System.Security.Cryptography.Hashing.Algorithms.xml",
-        "ref/dotnet/fr/System.Security.Cryptography.Hashing.Algorithms.xml",
-        "ref/dotnet/it/System.Security.Cryptography.Hash
```
Comment (DustinCampbell) [inline src/Compilers/VisualBasic/VbcCore/project.lock.json]: Intentionally removed?

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs [hunk 3]
```diff
@@ -0,0 +1,829 @@
...
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Loads the <see cref="SolutionInfo"/> for the specified solution file, including all projects referenced by the solution file and 
+        /// all the projects referenced by the project files.
+        /// </summary>
+        public async Task<SolutionInfo> LoadSolutionAsync(string solutionFilePath, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (DustinCampbell) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs]: Would LoadSolutionInfoAsync be a better name?

File: src/Compilers/VisualBasic/VbcCore/project.lock.json (hunk 1 above)
Comment (mattwar) [inline src/Compilers/VisualBasic/VbcCore/project.lock.json]: Honestly, I have no idea what that file is or how it got changed.

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs [hunk 4]
```diff
@@ -0,0 +1,829 @@
...
+            }
+
+            return absolutePath;
+        }
+
+        /// <summary>
+        /// Loads the <see cref="ProjectInfo"/> for the specified project. 
+        /// Referenced projects are converted to metadata references unless those projects are already loaded.
+        /// </summary>
+        public async Task<ProjectInfo> LoadProjectAsync(string projectFilePath, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (DustinCampbell) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs]: LoadProjectInfoAsync?

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs [hunk 5]
```diff
@@ -0,0 +1,829 @@
...
+            var loadedProjects = new Dictionary<ProjectId, ProjectInfo>();
+            var pid = await LoadProjectInfoAsync(projectFilePath, loadedProjects, preferMetadata: this.LoadMetadataForReferencedProjects, cancellationToken: cancellationToken).ConfigureAwait(false);
+            return loadedProjects[pid];
+        }
+
+        /// <summary>
+        /// Loads the <see cref="ProjectInfo"/>'s from the specified project file and all referenced projects.
+        /// </summary>
+        public Task<ProjectId> LoadProjectsAsync(string projectFilePath, Dictionary<ProjectId, ProjectInfo> loadedProjects, CancellationToken cancellationToken = default(CancellationToken))
+        {
```
Comment (DustinCampbell) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs]: This name doesn't feel quite right to me based on what it's doing. Also, it seems to need a lot of set up (Dictionary<ProjectId, ProjectInfo>). For what scenario would I use this API?

File: src/Workspaces/Core/Portable/PublicAPI.txt [hunk 6]
```diff
@@ -787,6 +787,7 @@ Microsoft.CodeAnalysis.Workspace.ScheduleTask<T>(System.Func<T> func, string tas
 Microsoft.CodeAnalysis.Workspace.Services.get -> Microsoft.CodeAnalysis.Host.HostWorkspaceServices
 Microsoft.CodeAnalysis.Workspace.SetCurrentSolution(Microsoft.CodeAnalysis.Solution solution) -> Microsoft.CodeAnalysis.Solution
 Microsoft.CodeAnalysis.Workspace.UnregisterText(Microsoft.CodeAnalysis.Text.SourceTextContainer textContainer) -> void
+Microsoft.CodeAnalysis.Workspace.UpdateReferencesAfterAdd() -> void
```
Comment (DustinCampbell) [inline src/Workspaces/Core/Portable/PublicAPI.txt]: It feels weird to see this exposed.

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs (hunk 3 above)
Comment (mattwar) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs]: Yep, probably.

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs (hunk 4 above)
Comment (mattwar) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs]: Yep.

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs (hunk 5 above)
Comment (mattwar) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs]: This was originally an internal API, but needs to be public now.  It basically needs to return all the ProjectInfo's for all referenced projects. But it also needs to return the ProjectId of the one associated with the project file you specified.  So it cannot just use SolutionInfo.  I don't have anything better yet.  I'm open to suggestions.

File: src/Workspaces/Core/Portable/PublicAPI.txt (hunk 6 above)
Comment (mattwar) [inline src/Workspaces/Core/Portable/PublicAPI.txt]: I was considering baking in a call to it directly, but there is not an ideal place for it to go.  It needs the host to make the decision.

**Comment by mattwar** [comment]
@dotnet-bot test this please

**Pull request merged**